File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> list = new ArrayList<>();
        sumHelper(candidates, 0, target, 0, new ArrayList<Integer>(), list);
        return list;
    }
    private void sumHelper(int[] arr, int i, int tar, int sum, List<Integer> res, List<List<Integer>> list){
        if(i>=arr.length){
            if(sum==tar){
                list.add(new ArrayList<>(res));
            }
            return;
        }
        if(sum>tar) return;
        res.add(arr[i]);
        sumHelper(arr, i, tar, sum+arr[i],res, list); //take
        res.remove(res.size()-1);
        sumHelper(arr, i+1, tar, sum, res, list);
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;
        csHelper(candidates, target, 0, temp, ans);
        return ans;
    }
    void csHelper(vector<int>& candidates, int target, int i, vector<int>& temp, vector<vector<int>>& ans){
        if(target<0) return;
        if(i>=candidates.size()){
            if(target!=0) return;
            vector<int> add = temp;
            ans.push_back(add);
            return;
        }
    
        // take or not take method with backtracking
        temp.push_back(candidates[i]);
        csHelper(candidates, target-candidates[i], i, temp, ans);
        temp.pop_back();
        csHelper(candidates, target, i+1, temp, ans);
    }
};